File: packages/core/src/utils/format.ts
/**
 * Text formatting utilities for RepoG.
 */

/**
 * Regex patterns for sensitive data that should be redacted.
 * - github_pat_: Fine-grained GitHub PATs
 * - ghp_: Classic GitHub PATs
 * - AIza: Gemini API keys
 */
const SENSITIVE_PATTERNS = [
  /github_pat_[A-Za-z0-9_]{20,}/g, // Fine-grained PATs
  /ghp_[A-Za-z0-9]{36,}/g, // Classic PATs
  /AIza[A-Za-z0-9_-]{35,}/g, // Gemini API keys
];

/**
 * Redact sensitive information from a string.
 *
 * Detects and replaces GitHub PATs (fine-grained and classic) and
 * Gemini API keys with [REDACTED] to prevent accidental exposure
 * in logs or error messages.
 *
 * @param str - The string to redact sensitive data from
 * @returns The string with sensitive data replaced by [REDACTED]
 */
export function redactSensitive(str: string): string {
  let result = str;
  for (const pattern of SENSITIVE_PATTERNS) {
    // Reset lastIndex for global regex
    pattern.lastIndex = 0;
    result = result.replace(pattern, '[REDACTED]');
  }
  return result;
}

/**
 * Options for word wrapping text.
 */
export interface WrapTextOptions {
  /** Maximum width for each line (default: 80) */
  width?: number;
  /** Indent string for continuation lines (default: '') */
  indent?: string;
  /** Indent string for the first line (default: '') */
  firstLineIndent?: string;
}

/**
 * Wrap text to a specified width with optional indentation.
 *
 * This function wraps text at word boundaries to fit within a maximum width.
 * Continuation lines can have a different indent than the first line.
 *
 * @param text - The text to wrap
 * @param options - Wrapping options
 * @returns The wrapped text with line breaks
 */
export function wrapText(text: string, options: WrapTextOptions = {}): string {
  const { width = 80, indent = '', firstLineIndent = '' } = options;

  // Normalize whitespace and split into words
  const words = text.replace(/\s+/g, ' ').trim().split(' ');

  if (words.length === 0 || (words.length === 1 && words[0] === '')) {
    return '';
  }

  const lines: string[] = [];
  let currentLine = firstLineIndent;
  let isFirstLine = true;

  for (const word of words) {
    const lineIndent = isFirstLine ? firstLineIndent : indent;
    const testLine = currentLine.length === lineIndent.length
      ? currentLine + word
      : currentLine + ' ' + word;

    if (testLine.length <= width) {
      currentLine = testLine;
    } else {
      // Line would exceed width
      if (currentLine.length > lineIndent.length) {
        // Current line has content, push it and start new line
        lines.push(currentLine);
        isFirstLine = false;
        currentLine = indent + word;
      } else {
        // Word is longer than available width, add it anyway
        currentLine = lineIndent + word;
      }
    }
  }

  // Push the last line if it has content
  if (currentLine.length > 0) {
    lines.push(currentLine);
  }

  return lines.join('\n');
}

/**
 * Format a star count for display with comma separators.
 *
 * @param stars - The number of stars
 * @returns Formatted star count string
 */
export function formatStars(stars: number): string {
  return stars.toLocaleString();
}

/**
 * Format a similarity score as a percentage.
 *
 * @param similarity - Similarity score between 0 and 1
 * @returns Formatted percentage string (e.g., "85.3%")
 */
export function formatSimilarity(similarity: number): string {
  return `${(similarity * 100).toFixed(1)}%`;
}

/**
 * Truncate text to a maximum length with ellipsis.
 *
 * @param text - The text to truncate
 * @param maxLength - Maximum length including ellipsis
 * @returns Truncated text
 */
export function truncateText(text: string, maxLength: number = 200): string {
  const cleaned = text.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
  if (cleaned.length <= maxLength) {
    return cleaned;
  }
  return cleaned.slice(0, maxLength - 3) + '...';
}

/**
 * Format a relative time string (e.g., "2 hours ago", "in 14 minutes").
 *
 * @param isoString - The ISO datetime string to format
 * @returns formatted relative time string
 */
export function formatRelativeTime(isoString: string): string {
  const date = new Date(isoString);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffSec = diffMs / 1000;

  if (diffMs >= 0) {
    // Past
    if (diffSec < 60) {
      return 'just now';
    }
    const minutes = Math.floor(diffSec / 60);
    if (minutes < 60) {
      return `${minutes} minute${minutes !== 1 ? 's' : ''} ago`;
    }
    const hours = Math.floor(minutes / 60);
    if (hours < 24) {
      return `${hours} hour${hours !== 1 ? 's' : ''} ago`;
    }
    const days = Math.floor(hours / 24);
    return `${days} day${days !== 1 ? 's' : ''} ago`;
  } else {
    // Future
    const absDiffSec = Math.abs(diffSec);
    const minutes = Math.ceil(absDiffSec / 60); // Round up for "in X minutes" so 59.1s becomes 1 min

    if (minutes < 60) {
      return `in ${minutes} minute${minutes !== 1 ? 's' : ''}`;
    }
    const hours = Math.floor(minutes / 60);
    if (hours < 24) {
      return `in ${hours} hour${hours !== 1 ? 's' : ''}`;
    }
    const days = Math.floor(hours / 24);
    return `in ${days} day${days !== 1 ? 's' : ''}`;
  }
}

File: internal/format/format.go
// Package format provides text formatting utilities for CLI output.
package format

import (
	"fmt"
	"os"
	"strings"
	"time"

	"golang.org/x/term"
)

// FormatStars formats a star count as "1.2k" for >= 1000, "342" for < 1000.
func FormatStars(count int) string {
	if count < 1000 {
		return fmt.Sprintf("%d", count)
	}
	if count < 10000 {
		return fmt.Sprintf("%.1fk", float64(count)/1000)
	}
	return fmt.Sprintf("%dk", count/1000)
}

// WrapText wraps text at maxWidth characters, indenting continuation lines
// with indent spaces.
func WrapText(text string, maxWidth, indent int) string {
	// Normalize whitespace
	text = strings.Join(strings.Fields(text), " ")
	if text == "" {
		return ""
	}

	words := strings.Split(text, " ")
	var lines []string
	var currentLine strings.Builder
	indentStr := strings.Repeat(" ", indent)

	isFirstLine := true
	for _, word := range words {
		lineLen := currentLine.Len()
		wordLen := len(word)

		// Calculate effective width for this line
		effectiveWidth := maxWidth
		if !isFirstLine {
			effectiveWidth -= indent
		}

		// Check if adding this word would exceed width
		if lineLen > 0 && lineLen+1+wordLen > effectiveWidth {
			lines = append(lines, currentLine.String())
			currentLine.Reset()
			isFirstLine = false
		}

		if currentLine.Len() > 0 {
			currentLine.WriteString(" ")
		}
		currentLine.WriteString(word)
	}

	if currentLine.Len() > 0 {
		lines = append(lines, currentLine.String())
	}

	// Add indentation to continuation lines
	for i := 1; i < len(lines); i++ {
		lines[i] = indentStr + lines[i]
	}

	return strings.Join(lines, "\n")
}

// Separator returns a string of '─' characters at the given width (max 60).
func Separator(width int) string {
	if width > 60 {
		width = 60
	}
	if width < 1 {
		width = 1
	}
	return strings.Repeat("─", width)
}

// TerminalWidth returns the current terminal width, defaulting to 80.
func TerminalWidth() int {
	width, _, err := term.GetSize(int(os.Stdout.Fd()))
	if err != nil || width <= 0 {
		return 80
	}
	return width
}

// FormatSimilarity formats a similarity score as a percentage.
func FormatSimilarity(similarity float64) string {
	return fmt.Sprintf("%.1f%%", similarity*100)
}

// TruncateText truncates text to a maximum length with ellipsis.
func TruncateText(text string, maxLength int) string {
	// Normalize whitespace
	cleaned := strings.Join(strings.Fields(text), " ")
	if len(cleaned) <= maxLength {
		return cleaned
	}
	if maxLength <= 3 {
		return cleaned[:maxLength]
	}
	return cleaned[:maxLength-3] + "..."
}

// FormatRelativeTime formats a relative time string (e.g., "2 hours ago", "in 14 minutes").
func FormatRelativeTime(isoString string) string {
	t, err := time.Parse(time.RFC3339, isoString)
	if err != nil {
		// Try other common formats
		t, err = time.Parse("2006-01-02T15:04:05Z", isoString)
		if err != nil {
			return isoString
		}
	}

	now := time.Now()
	diff := now.Sub(t)

	if diff >= 0 {
		// Past
		seconds := int(diff.Seconds())
		if seconds < 60 {
			return "just now"
		}

		minutes := seconds / 60
		if minutes < 60 {
			if minutes == 1 {
				return "1 minute ago"
			}
			return fmt.Sprintf("%d minutes ago", minutes)
		}

		hours := minutes / 60
		if hours < 24 {
			if hours == 1 {
				return "1 hour ago"
			}
			return fmt.Sprintf("%d hours ago", hours)
		}

		days := hours / 24
		if days == 1 {
			return "1 day ago"
		}
		return fmt.Sprintf("%d days ago", days)
	}

	// Future
	absDiff := -diff
	seconds := int(absDiff.Seconds())

	minutes := (seconds + 59) / 60 // Round up
	if minutes < 60 {
		if minutes == 1 {
			return "in 1 minute"
		}
		return fmt.Sprintf("in %d minutes", minutes)
	}

	hours := minutes / 60
	if hours < 24 {
		if hours == 1 {
			return "in 1 hour"
		}
		return fmt.Sprintf("in %d hours", hours)
	}

	days := hours / 24
	if days == 1 {
		return "in 1 day"
	}
	return fmt.Sprintf("in %d days", days)
}

// RedactSensitive redacts sensitive information from a string.
func RedactSensitive(str string) string {
	// Patterns for sensitive data
	patterns := []string{
		"github_pat_",
		"ghp_",
		"AIza",
	}

	result := str
	for _, pattern := range patterns {
		for {
			idx := strings.Index(result, pattern)
			if idx == -1 {
				break
			}
			// Find the end of the token (whitespace or end of string)
			end := idx + len(pattern)
			for end < len(result) && !isWhitespace(result[end]) {
				end++
			}
			result = result[:idx] + "[REDACTED]" + result[end:]
		}
	}

	return result
}

func isWhitespace(c byte) bool {
	return c == ' ' || c == '\t' || c == '\n' || c == '\r'
}
